### PR TITLE
Adds new entities and updates existing ones

### PR DESCRIPTION
**Pull request by randomdavis**
Passes all tests on AppHarbor. Check our Google Drive for the diagram of the new schema.

### CODE REVIEW

File: AES.Entities/Tables/ApplicantUser.cs [hunk 1]
```diff
@@ -27,5 +27,11 @@ public class ApplicantUser
         public DateTime DOB { get; set; }
 
         public virtual UserInfo UserInfo { get; set; }
+
+        // Only one per applicant
+        public virtual Availability Availability { get; set; }
+
+        public DateTime? CallStartTime { get; set; }
```
Comment (DavidA94) [inline AES.Entities/Tables/ApplicantUser.cs]: Why can these be null?

File: AES.Entities/Tables/Application.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace AES.Entities.Tables
+{
+    public class Application
+    {
+        [Key]
+        [Required]
+        public int ID { get; set; }
+
+        // Which job the application is for
```
Comment (DavidA94) [inline AES.Entities/Tables/Application.cs]: Instead of a single line comment, type three slashes, which will create a < summary> section. Anything between the start and close tags will be visible when somebody uses the properties. Repeat for similar property explanations.

File: AES.Entities/Tables/ApplicationMultiAnswer.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace AES.Entities.Tables
+{
+    class ApplicationMultiAnswer
+    {
+        [Key]
+        [Required]
+        public int ID { get; set; }
+
+        // Which question this is answering
```
Comment (DavidA94) [inline AES.Entities/Tables/ApplicationMultiAnswer.cs]: Ditto on < summary> tags.

File: AES.Entities/Tables/ApplicationShortAnswer.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace AES.Entities.Tables
+{
+    class ApplicationShortAnswer
+    {
+        [Key]
+        [Required]
+        public int ID { get; set; }
+
+        // Which question this is answering
```
Comment (DavidA94) [inline AES.Entities/Tables/ApplicationShortAnswer.cs]: Ditto on < summary> tags.

File: AES.Entities/Tables/Availability.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace AES.Entities.Tables
+{
+    public class Availability
+    {
+        [Key]
+        [Required]
+        public int ID { get; set; }
+
+        public DateTime? SundayStart { get; set; }
```
Comment (DavidA94) [inline AES.Entities/Tables/Availability.cs:12]: Why null?

File: AES.Entities/Tables/EducationHistory.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        [Required]
+        [StringLength(128)]
+        public string SchoolName { get; set; }
+
+        // Not required as the applicant may not know
+        [StringLength(128)]
+        public string SchoolAddress { get; set; }
+
+        [Required]
+        [StringLength(128)]
```
Comment (DavidA94) [inline AES.Entities/Tables/EducationHistory.cs]: This can be the same length as the city for the applicants address.

File: AES.Entities/Tables/EducationHistory.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        public string SchoolCity { get; set; }
+
+        // Not required as the applicant may not know or it may not be relevant
+        [StringLength(64)]
+        public string SchoolState { get; set; }
```
Comment (DavidA94) [inline AES.Entities/Tables/EducationHistory.cs:33]: Never seen a 64 character state. :) Make length 2.

File: AES.Entities/Tables/EducationHistory.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+        // Not required as the applicant may not know or it may not be relevant
+        [StringLength(16)]
```
Comment (DavidA94) [inline AES.Entities/Tables/EducationHistory.cs]: Should be an integer with min 0, max 99999.

File: AES.Entities/Tables/EducationHistory.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        public string SchoolZip { get; set; }
+
+        [Required]
+        [StringLength(128)]
+        public string SchoolCountry { get; set; }
```
Comment (DavidA94) [inline AES.Entities/Tables/EducationHistory.cs:43]: If you want to allow counties, I'd say 32 characters would be plenty. "Republic of Czechoslovakia" is only 26 characters.

File: AES.Entities/Tables/ApplicantUser.cs (hunk 1 above)
Comment (randomdavis) [inline AES.Entities/Tables/ApplicantUser.cs]: Otherwise the database throws an overflow error

File: AES.Entities/Tables/Application.cs (hunk 2 above)
Comment (randomdavis) [inline AES.Entities/Tables/Application.cs]: Will do

File: AES.Entities/Tables/ApplicationMultiAnswer.cs (hunk 3 above)
Comment (randomdavis) [inline AES.Entities/Tables/ApplicationMultiAnswer.cs]: Will do

File: AES.Entities/Tables/ApplicationShortAnswer.cs (hunk 4 above)
Comment (randomdavis) [inline AES.Entities/Tables/ApplicationShortAnswer.cs]: Will do

File: AES.Entities/Tables/Availability.cs (hunk 5 above)
Comment (randomdavis) [inline AES.Entities/Tables/Availability.cs:12]: Otherwise null dates will break the database

File: AES.Entities/Tables/EducationHistory.cs (hunk 6 above)
Comment (randomdavis) [inline AES.Entities/Tables/EducationHistory.cs]: Will do

File: AES.Entities/Tables/EducationHistory.cs (hunk 7 above)
Comment (randomdavis) [inline AES.Entities/Tables/EducationHistory.cs:33]: Will do

File: AES.Entities/Tables/EducationHistory.cs [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+        // Not required as the applicant may not know
+        [StringLength(32)]
+        public string SchoolPhone { get; set; }
```
Comment (DavidA94) [inline AES.Entities/Tables/EducationHistory.cs:46]: Thinking 16 should be fine for this, assuming international numbers. But if you want, just drop international support.

File: AES.Entities/Tables/EducationHistory.cs (hunk 8 above)
Comment (randomdavis) [inline AES.Entities/Tables/EducationHistory.cs]: Will do (I was thinking of Canadian ZIP codes which can be numbers and letters)

File: AES.Entities/Tables/EducationHistory.cs [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+        [Required]
+        public double YearsAddented { get; set; }
+
+        // Can be used to determine if the applicant has graduated or will graduate (it would be a future date)
+        public DateTime GraduationDate { get; set; }
+
+        [Required]
+        [StringLength(128)]
```
Comment (DavidA94) [inline AES.Entities/Tables/EducationHistory.cs]: 64 for this one and below should be plenty.

File: AES.Entities/Tables/JobHistory.cs [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        [Key]
+        [Required]
+        public int ID { get; set; }
+
+        [Required]
+        public ApplicantUser Applicant { get; set; }
+
+        [Required]
+        [StringLength(128)]
```
Comment (DavidA94) [inline AES.Entities/Tables/JobHistory.cs:19]: Ditto on the stuff for the address.

File: AES.Entities/Tables/JobHistory.cs [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        [StringLength(32)]
+        public string EmployerPhone { get; set; }
+
+        [Required]
+        [StringLength(128)]
+        public string SupervisorName { get; set; }
+
+        // Not Required
+        [StringLength(128)]
+        public string StartingSalary { get; set; }
```
Comment (DavidA94) [inline AES.Entities/Tables/JobHistory.cs]: Use type decimal, not string.

File: AES.Entities/Tables/JobHistory.cs [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+        // Not Required
+        [StringLength(128)]
+        public string EndingSalary { get; set; }
```
Comment (DavidA94) [inline AES.Entities/Tables/JobHistory.cs]: Ditto

File: AES.Entities/Tables/JobQuestion.cs [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        [Required]
+        public int ID { get; set; }
+
+        // Which job this question is for
+        [Required]
+        public Job Job { get; set; }
+
+        // 1 for short answer, 2 for radio multi choice, 3 for checkbox multi choice
+        [Required]
+        [Range(1, 3)]
```
Comment (DavidA94) [inline AES.Entities/Tables/JobQuestion.cs]: Use an enum.

File: AES.Entities/Tables/UserInfo.cs [hunk 16]
```diff
@@ -31,6 +26,6 @@ public class UserInfo
         [RegularExpression(@"\d{3}-\d{3}-\d{4}")]
         public string Phone { get; set; }
 
-        public decimal SalaryExpectation { get; set; }
+        public double SalaryExpectation { get; set; }
```
Comment (DavidA94) [inline AES.Entities/Tables/UserInfo.cs]: Nope, decimal is right. That's a C# special type for money. We don't need the precision a double gives.

File: AES.SecuritySvc.Tests/SecurityUnitTest.cs [hunk 17]
```diff
@@ -14,12 +14,12 @@ public class SecurityUnitTest
...
         private const string SSN = "11-22-3344";
         private readonly string SSN_CRYPT;
-        private readonly DateTime DOB;
+        private readonly DateTime TEST_DOB;
 
         public SecurityUnitTest()
         {
             SSN_CRYPT = Encryption.Encrypt(SSN);
-            DOB = new DateTime(1970, 1, 1);
+            TEST_DOB = new DateTime(1964, 6, 2);
```
Comment (DavidA94) [inline AES.SecuritySvc.Tests/SecurityUnitTest.cs]: Does this work? Usually 1970 is the earliest a date can go. Why the change with the name as well?

File: AES.Entities/Tables/ApplicantUser.cs (hunk 1 above)
Comment (DavidA94) [inline AES.Entities/Tables/ApplicantUser.cs]: We should handle that on the code side, but those DateTime's should be required.

File: AES.Entities/Tables/EducationHistory.cs (hunk 8 above)
Comment (DavidA94) [inline AES.Entities/Tables/EducationHistory.cs]: Fair enough. Up to 6 characters then if you want to have international support, and put a comment why it's a string.

File: AES.Entities/Tables/ApplicantUser.cs (hunk 1 above)
Comment (randomdavis) [inline AES.Entities/Tables/ApplicantUser.cs]: Then I'll have to update your test to include that, since otherwise it will crash

File: AES.Entities/Tables/EducationHistory.cs (hunk 8 above)
Comment (randomdavis) [inline AES.Entities/Tables/EducationHistory.cs]: I'll just go with what you already had

File: AES.Entities/Tables/EducationHistory.cs (hunk 9 above)
Comment (randomdavis) [inline AES.Entities/Tables/EducationHistory.cs:43]: I'll do 32

File: AES.Entities/Tables/EducationHistory.cs (hunk 10 above)
Comment (randomdavis) [inline AES.Entities/Tables/EducationHistory.cs:46]: I'm just going with what you already did

File: AES.Entities/Tables/JobHistory.cs (hunk 12 above)
Comment (randomdavis) [inline AES.Entities/Tables/JobHistory.cs:19]: Will do

File: AES.Entities/Tables/JobHistory.cs (hunk 13 above)
Comment (randomdavis) [inline AES.Entities/Tables/JobHistory.cs]: Will do

File: AES.Entities/Tables/JobHistory.cs (hunk 14 above)
Comment (randomdavis) [inline AES.Entities/Tables/JobHistory.cs]: Will do

File: AES.Entities/Tables/JobQuestion.cs (hunk 15 above)
Comment (randomdavis) [inline AES.Entities/Tables/JobQuestion.cs]: I will figure that out and try to do that

File: AES.Entities/Tables/UserInfo.cs (hunk 16 above)
Comment (randomdavis) [inline AES.Entities/Tables/UserInfo.cs]: Fixed

File: AES.SecuritySvc.Tests/SecurityUnitTest.cs (hunk 17 above)
Comment (randomdavis) [inline AES.SecuritySvc.Tests/SecurityUnitTest.cs]: 1970 is the earliest a timestamp can go, but historic dates can go back much further

File: AES.Entities/Tables/JobQuestion.cs [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+using System.ComponentModel.DataAnnotations;
+
+namespace AES.Entities.Tables
+{
+    class JobQuestion
+    {
+
+        public enum QuestionType
+        {
+            SHORT = 1,
```
Comment (DavidA94) [inline AES.Entities/Tables/JobQuestion.cs:11]: Starting value not needed. Default of zero will be fine.

File: AES.Entities/Tables/Availability.cs (hunk 5 above)
Comment (DavidA94) [inline AES.Entities/Tables/Availability.cs:12]: Should have updated these too to be non-null.
